### PR TITLE
Neighbourhood diagonal

### PR DESCRIPTION
**Pull request by niravshah241**
Once the domain is divided across different MPI ranks, the rank corresponding to the neighbour of a given subdomain in a given direction needs to be identified. Currently the functionality is implemented to identify neighbours in TOP, RIGHT, BOTTOM and LEFT direction which share a common edge with the given subdomain. This Draft PR aims to add functionality to identify neighbours in TOP_RIGHT, BOTTOM_RIGHT, BOTTOM_LEFT and TOP_LEFT corners.

Specifically, the changes include:

- Addition of [is_diagonal_neighbour](https://github.com/nextsimhub/domain_decomp/blob/neighbourhood_diagonal/Partitioner.cpp#L55)
- Addition of [domain_overlap_diagonal](https://github.com/nextsimhub/domain_decomp/blob/neighbourhood_diagonal/DomainUtils.cpp#L37)
- Addition of [Vertex](https://github.com/nextsimhub/domain_decomp/blob/neighbourhood_diagonal/DomainUtils.hpp#L16) for corner neighbours
- Addition of [halo_start_neighbour](https://github.com/nextsimhub/domain_decomp/blob/neighbourhood_diagonal/Partitioner.cpp#L116)
- Addition of [discover_diagonal_neighbours](https://github.com/nextsimhub/domain_decomp/blob/neighbourhood_diagonal/Partitioner.cpp#L493)

This PR has merged the following PRs:
- #81
- #78 
- #77 

It now contains all the functionality required.

### CODE REVIEW

**Comment by TomMelt** [comment]
@niravshah241 thanks for drafting this PR. I know that it is draft stage but draft PRs still shouldn't contain CMake build files. Please remove these as discussed last meeting. Please only commit changes for relevant source files.

File: DomainUtils.cpp [hunk 1]
```diff
@@ -33,3 +33,33 @@ int domain_overlap(const Domain d1, const Domain d2, const Edge edge)
     }
     return overlap;
 }
+
+bool domain_overlap_diagonal(const Domain d1, const Domain d2, const Vertex vertex)
```
Comment (TomMelt) [inline DomainUtils.cpp]: As discussed last meeting, this function `domain_overlap_diagonal` is not needed for the corner neighbours. It can only ever be a fixed size i.e., 1. Please remove this.

File: Partitioner.cpp [hunk 2]
```diff
@@ -433,3 +489,57 @@ void Partitioner::discover_neighbours()
         }
     }
 }
+
+void Partitioner::discover_diagonal_neighbours(){
```
Comment (TomMelt) [inline Partitioner.cpp]: Could you please rename this function to `discover_corner_neighbours`? This will be clearer.

File: Partitioner.hpp [hunk 3]
```diff
@@ -234,6 +237,7 @@ class LIB_EXPORT Partitioner {
      * @return starting index of halo for the flattened domain array
      */
     int halo_start(const Domain d1, const Domain d2, const Edge edge);
+    int halo_start_neighbour(const Domain d1, const Domain d2, const Vertex vertex);
```
Comment (TomMelt) [inline Partitioner.hpp]: I would also consider renaming this to `halo_start_corner`. We can chat in our next catch up but I think we might not need this in the end either.

File: spack.yaml
Comment (TomMelt) [inline spack.yaml:1]: We should look at this `spack.yaml` in a separate pull request

File: spack_trial.yaml
Comment (TomMelt) [inline spack_trial.yaml:1]: I dont think this should be uploaded to the PR. It looks the same as the other `spack.yaml` file

File: DomainUtils.cpp (hunk 1 above)
Comment (niravshah241) [reply DomainUtils.cpp]: Removed.

File: spack.yaml
Comment (niravshah241) [reply spack.yaml:1]: I removed this spack.yaml from this PR. I had used spack from NextsimDG. It had necessary changes incorporated (see PR https://github.com/nextsimhub/nextsimdg/pull/908) .

File: spack_trial.yaml
Comment (niravshah241) [reply spack_trial.yaml:1]: Removed

File: Partitioner.cpp [hunk 4]
```diff
@@ -52,6 +52,40 @@ bool Partitioner::is_neighbour(
     }
 }
 
+bool Partitioner::is_corner_neighbour(
```
Comment (dorchard) [inline Partitioner.cpp]: Add some documentation here to explain how this works?

File: Partitioner.cpp [hunk 5]
```diff
@@ -52,6 +52,40 @@ bool Partitioner::is_neighbour(
...
+    const Domain d1, const Domain d2, const Vertex vertex, const bool is_px, const bool is_py)
+{
+    if (vertex == TOP_LEFT) {
+        // is_px and is_py is not updatd ath moment as periodic boundaries are not considered.
```
Comment (dorchard) [inline Partitioner.cpp]: Typo here.

File: Partitioner.cpp [hunk 6]
```diff
@@ -79,6 +113,40 @@ int Partitioner::halo_start(const Domain d1, const Domain d2, const Edge edge)
     return start;
 }
 
+int Partitioner::halo_corner_start(const Domain d1, const Domain d2, const Vertex vertex)
```
Comment (dorchard) [inline Partitioner.cpp]: Make clear that a pre-condition of using this function is that `is_corner_neighbour` is true.

File: Partitioner.cpp [hunk 7]
```diff
@@ -252,6 +349,21 @@ void Partitioner::save_metadata(const std::string& filename) const
...
 
+    int num_corner_vid[N_VERTEX];
+    int ids_corner_vid[N_VERTEX];
+    int halo_corner_starts_vid[N_VERTEX];
+    for (auto vertex : vertices) {
+        // Connectivity group
+        NC_CHECK(nc_def_var(connectivity_gid, (corner_dir_names[vertex] + "_neighbours").c_str(), NC_INT, 1,
+            &corner_dimid, &num_corner_vid[vertex]));
+        // NC_CHECK(nc_def_var(connectivity_gid, (corner_dir_names[vertex] + "_neighbour_ids").c_str(), NC_INT,
+        //     1, &corner_dimids[vertex], &ids_corner_vid[vertex]));
```
Comment (dorchard) [inline Partitioner.cpp]: As discussed, cleanup this comment (and others).

File: Partitioner.cpp [hunk 8]
```diff
@@ -52,6 +52,40 @@ bool Partitioner::is_neighbour(
     }
 }
 
+bool Partitioner::is_corner_neighbour(
+    const Domain d1, const Domain d2, const Vertex vertex, const bool is_px, const bool is_py)
+{
+    if (vertex == TOP_LEFT) {
+        // is_px and is_py is not updatd ath moment as periodic boundaries are not considered.
+        // Check if TOP Left neighbour i.e.,
```
Comment (dorchard) [inline Partitioner.cpp]: Some documentation on what `is_px` and `is_py` are for is needed.
I don't understand why the TOP_LEFT case doesn't look at either `is_py` or `is_px` but all the other cases do... why would something be special here?

File: DomainUtils.cpp
Comment (TomMelt) [inline DomainUtils.cpp:1]: @niravshah241 as a rule, I would generally just remove this change. It's just formatting. Unless it's required by the linter (which I dont think it is).

File: Partitioner.cpp (hunk 2 above)
Comment (niravshah241) [reply Partitioner.cpp]: Removed this function completely and added logic to `discover_neighbours()`.

File: Partitioner.hpp (hunk 3 above)
Comment (niravshah241) [reply Partitioner.hpp]: Renamed it to `halo_corner_start`.

**Comment by niravshah241** [comment]
Pending tasks:

- [x] Remove minor formatting related errors
- [ ] ~~Add one more test (say test_5) with overlap with other domain on top and bottom edge~~ To be addressed as separately, if necessary
- [x] Rebasing this branch onto main
- [x] Change n,m to x,y in the tests
- [x] Add a comment to the nested functions saying something like, "this should only be called if is_neighbour is true
- [x] Periodic cases (PR https://github.com/nextsimhub/domain_decomp/pull/77)
- [x] Add [Link](https://nextsim-dg.readthedocs.io/en/latest/halo-exchange.html) of Halo exchange logic
- [ ] ~~Unit test for ```is_corner_neighbour```~~ To be addressed as separately and not in this PR

File: Partitioner.cpp [hunk 9]
```diff
@@ -213,6 +302,14 @@ void Partitioner::save_metadata(const std::string& filename) const
...
     }
 
+    // Define dimensions in netCDF file
+    int corner_dimid;
+    std::vector<int> corner_dimids(N_VERTEX);
+    NC_CHECK(nc_def_dim(nc_id, "corner_P", _total_num_procs, &corner_dimid));
+    for (auto vertex : vertices) {
+        NC_CHECK(nc_def_dim(nc_id, corner_dir_chars[vertex].c_str(), corner_dims[vertex], &corner_dimids[vertex]));
+    }
+
```
Comment (TomMelt) [inline Partitioner.cpp]: @niravshah241 you dont need to define `corner_p`/`corner_dimid` because the number of processes is the same whether we have edges or corners. So just use the existing `dimid` variable instead

Comment (TomMelt) [reply Partitioner.cpp]: FYI I came across this whilst adding periodic neighbours so I have fixed it in my PR #77 .

see e.g., https://github.com/nextsimhub/domain_decomp/blob/bef6fc7e1d781d945129a631450a0fcc94d1a700/Partitioner.cpp#L458-L459

**Comment by TomMelt** [comment]
This PR has merged the following PRs:
- #81
- #78 
- #77 

It now contains all the functionality required.

**Review by TomMelt — APPROVED** [review]
All ready to go now.
